scripts/helpers.js

```diff
@@ -0,0 +1,137 @@
+const { spawn } = childProcess;
+
+const EXPECTED_LOAD_ERRORS = /ECONNREFUSED|ECONNRESET|did not get to load all resources/;
+
+function spawnPiped(cmd, options) {
+  function pipeWithPrefix(stream, prefix = '') {
+    return data => {
+      const prefixedMsg = data
+        .toString('utf8')
+        .trimRight()
+        .replace(/^/gm, prefix);
+      stream.write(`${prefixedMsg}\n`, 'utf8');
+    };
+  }
+
+  const child = spawn(cmd[0], cmd.slice(1), {
+    stdio: 'pipe',
+    shell: true,
+    ...options,
+  });
+  child.on('error', handleError);
+  child.on('exit', code => {
+    if (code) {
+      print(`Child '${cmd.join(' ')}' exited with code ${code}`);
+      process.exit(code);
+    }
+  });
+  child.stdout.on(
+    'data',
+    pipeWithPrefix(process.stdout, `[${cmd.join(' ')}].out: `),
+  );
+  child.stderr.on(
+    'data',
+    pipeWithPrefix(process.stderr, `[${cmd.join(' ')}].err: `),
+  );
+
+  return child;
+}
+
+function handleError(err) {
+  process.stdout.write(`${err.name}: ${err.stack || err.message}\n`);
+  if (typeof err.code === 'number') {
+    process.exit(err.code);
+  } else {
+    process.exit(1);
+  }
+}
+
+function waitFor(fn) {
+  return new Promise(resolve => {
+    const handle = setInterval(() => {
+      if (fn()) {
+        clearInterval(handle);
+        resolve();
+        return;
+      }
+    }, 100);
+  });
+}
+
+async function waitForExit(child) {
+  if (child.exitCode !== null) {
+    throw new Error(`Child already exited with code ${child.exitCode}`);
+  }
+  await new Promise((resolve, reject) =>
+    child.once('exit', code => {
+      if (code) {
+        reject(new Error(`Child exited with code ${code}`));
+      } else {
+        print('Child finished');
+        resolve();
+      }
+    }),
+  );
+}
+
+async function waitForPageWithText(
+  browser,
+  path,
+  text,
+  { intervalMs = 1000, maxAttempts = 240 } = {},
+) {
+  let attempts = 0;
+  for (;;) {
+    try {
+      await new Promise(resolve => setTimeout(resolve, intervalMs));
+      await browser.visit(path);
+      break;
+    } catch (error) {
+      if (error.message.match(EXPECTED_LOAD_ERRORS)) {
+        attempts++;
+        if (attempts > maxAttempts) {
+          throw new Error(
+            `Failed to load page '${path}', max number of attempts reached`,
+          );
+        }
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  const escapedText = text.replace(/"/g, '\\"');
+  browser.assert.evaluate(
+    `Array.from(document.querySelectorAll("*")).some(el => el.textContent === "${escapedText}")`,
+    true,
+    `expected to find text ${text}`,
+  );
+}
+
+function print(msg) {
+  return process.stdout.write(`${msg}\n`);
+}
+
+module.exports.spawnPiped = spawnPiped;
+module.exports.handleError = handleError;
+module.exports.waitFor = waitFor;
+module.exports.waitForExit = waitForExit;
+module.exports.waitForPageWithText = waitForPageWithText;
+module.exports.print = print;
```

Comment:
is it possible to do `module.exports = { spawnPiped, handleError, etcetc }`?

Reply:
👍